pkg/image/registry/imagestream/etcd/image.go

```diff
@@ -0,0 +1,159 @@
+	"k8s.io/apimachinery/pkg/watch"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/client-go/tools/cache"
+
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+)
+
+type ImageLayerIndex interface {
+	cache.SharedIndexInformer
+}
+
+type ImageStore interface {
+	rest.Watcher
+	rest.Lister
+}
+
+func NewImageLayerIndex(store ImageStore) ImageLayerIndex {
+	ctx := apirequest.NewContext()
+	informer := cache.NewSharedIndexInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			obj, err := store.List(ctx, &metainternalversion.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+			})
+			if err != nil {
+				return nil, err
+			}
+			list, ok := obj.(*imageapi.ImageList)
+			if !ok {
+				return nil, fmt.Errorf("unexpected store type %T for layer index", obj)
+			}
+			out := &metainternalversion.List{
+				Items: make([]runtime.Object, len(list.Items)),
+			}
+			for i, image := range list.Items {
+				out.Items[i] = &ImageLayers{
+					Name:     image.Name,
+					Layers:   image.DockerImageLayers,
+					Manifest: manifestFromImage(&image),
+				}
+			}
+			return out, nil
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			w, err := store.Watch(ctx, &metainternalversion.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+			})
+			if err != nil {
+				return nil, err
+			}
+			return watch.Filter(w, func(in watch.Event) (out watch.Event, keep bool) {
+				if in.Object == nil {
+					return in, true
+				}
+				image, ok := in.Object.(*imageapi.Image)
+				if !ok {
+					return in, true
+				}
+				in.Object = &ImageLayers{
+					Name:     image.Name,
+					Layers:   image.DockerImageLayers,
+					Manifest: manifestFromImage(image),
+				}
+				return in, true
+			}), nil
+		},
+	}, &ImageLayers{}, 0, cache.Indexers{
+		"layers": func(obj interface{}) ([]string, error) {
+			entry, ok := obj.(*ImageLayers)
+			if !ok {
+				return nil, fmt.Errorf("unexpected cache object %T", obj)
+			}
+			keys := make([]string, 0, len(entry.Layers))
+			for _, layer := range entry.Layers {
+				keys = append(keys, layer.Name)
+			}
+			return keys, nil
+		},
+	})
+	return informer
+}
+
+// manifestFromImage attempts to find a manifest blob description from
+// an image. Images older than schema2 in Docker do not have a manifest blob.
+func manifestFromImage(image *imageapi.Image) *imageapi.ImageLayer {
+	if image.DockerImageManifestMediaType != "application/vnd.docker.distribution.manifest.v2+json" {
+		return nil
+	}
+	return &imageapi.ImageLayer{
+		Name:      image.DockerImageMetadata.ID,
+		MediaType: image.DockerImageManifestMediaType,
+	}
+}
+
+type ImageLayers struct {
+	Name     string
+	Manifest *imageapi.ImageLayer
+	Layers   []imageapi.ImageLayer
+}
+
+var _ metav1.Object = &ImageLayers{}
+
+func (l *ImageLayers) GetObjectKind() schema.ObjectKind { return &metav1.TypeMeta{} }
```

Comment:
looks like we are forcing `ImageLayers` to be an `runtime.Object`. At the very least this neads a very good reason documented as a comment above.

Reply:
@smarterclayton why not making this an actual API object?

Reply:
This is a minimal cache.  Cache doesn't actually need runtime object except there are bugs in cache.  I'm going to correct those, which is why there is a comment.  Informers are supposed to depend on list-watch (which requires resource version) but the cache key is arbitrary.